overlord/snapstate/handlers_link_test.go

```diff
@@ -1397,6 +1423,213 @@ func (s *linkSnapSuite) TestDoLinkSnapdRemovesAppArmorProfilesOnSnapdDowngrade(c
+	s.state.Lock()
+	t := s.state.NewTask("link-snap", "test")
+	sinfo := snap.SideInfo{
+		RealName: "foo",
+		Revision: snap.R(33),
+		SnapID:   "foo-id",
+	}
+	t.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: &sinfo,
+		Channel:  "beta",
+		UserID:   2,
+	})
+	s.state.NewChange("sample", "...").AddTask(t)
+
+	s.state.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// link-snap task should detect that we don't have upcoming
+	// uninhibit-snap tasks in the current change and call
+	// backend.UninhibitSnap itself
+	expected := fakeOps{
+		{
+			op:    "candidate",
+			sinfo: sinfo,
+		},
+		{
+			op:   "link-snap",
+			path: filepath.Join(dirs.SnapMountDir, "foo/33"),
+		},
+		{
+			op:   "uninhibit-snap",
+			name: "foo",
+		},
+	}
+	c.Check(s.fakeBackend.ops.Ops(), DeepEquals, expected.Ops())
+	c.Check(s.fakeBackend.ops, DeepEquals, expected)
+}
+
+func (s *linkSnapSuite) TestDoLinkSnapMaybeUninhibitSnapOffNoraml(c *C) {
+	s.state.Lock()
+
+	sinfo := snap.SideInfo{
+		RealName: "foo",
+		Revision: snap.R(33),
+		SnapID:   "foo-id",
+	}
+	snapsup := snapstate.SnapSetup{
+		SideInfo: &sinfo,
+		Channel:  "beta",
+		UserID:   2,
+	}
+
+	t1 := s.state.NewTask("link-snap", "test")
+	t1.Set("snap-setup", &snapsup)
+	t2 := s.state.NewTask("uninhibit-snap", "test")
+	t2.WaitFor(t1)
+	t2.Set("snap-setup", &snapsup)
+
+	chg := s.state.NewChange("sample", "...")
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+
+	s.state.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	// link-snap task should detect that we don't have upcoming
+	// uninhibit-snap tasks in the current change and call
+	// backend.UninhibitSnap itself
+	expected := fakeOps{
+		{
+			op:    "candidate",
+			sinfo: sinfo,
+		},
+		{
+			op:   "link-snap",
+			path: filepath.Join(dirs.SnapMountDir, "foo/33"),
+		},
+	}
+	c.Check(s.fakeBackend.ops.Ops(), DeepEquals, expected.Ops())
+	c.Check(s.fakeBackend.ops, DeepEquals, expected)
+
+	// uninhibit-snap task should run now
+	s.se.Ensure()
+	s.se.Wait()
+
+	expected = fakeOps{
+		{
+			op:    "candidate",
+			sinfo: sinfo,
+		},
+		{
+			op:   "link-snap",
+			path: filepath.Join(dirs.SnapMountDir, "foo/33"),
+		},
+		{
+			op:   "uninhibit-snap",
+			name: "foo",
+		},
+	}
+	c.Check(s.fakeBackend.ops.Ops(), DeepEquals, expected.Ops())
+	c.Check(s.fakeBackend.ops, DeepEquals, expected)
+}
+
+func (s *linkSnapSuite) TestDoLinkSnapMaybeUninhibitSnapOffAnotherChange(c *C) {
+	s.state.Lock()
+
+	sinfo := snap.SideInfo{
+		RealName: "foo",
+		Revision: snap.R(33),
+		SnapID:   "foo-id",
+	}
+	snapsup := snapstate.SnapSetup{
+		SideInfo: &sinfo,
+		Channel:  "beta",
+		UserID:   2,
+	}
+
+	t1 := s.state.NewTask("link-snap", "test")
+	t1.Set("snap-setup", &snapsup)
+	s.state.NewChange("sample-1", "...").AddTask(t1)
+
+	t2 := s.state.NewTask("uninhibit-snap", "test")
+	t2.Set("snap-setup", &snapsup)
+	s.state.NewChange("sample-2", "...").AddTask(t2)
+
+	s.state.Unlock()
+
+	for i := 0; i < 3; i++ {
```

Comment:
Why is the loop necessary?

Reply:
you are right this is not needed, thank you for noting